File: crates/a3s-ahp/2.3.1/examples/websocket_server.ts
#!/usr/bin/env node
/**
 * AHP WebSocket Server (TypeScript)
 *
 * A WebSocket-based AHP harness server with bidirectional streaming support.
 *
 * Features:
 * - Persistent connections with low latency
 * - Concurrent client support
 * - Ping/Pong heartbeat
 * - Batch processing
 * - Depth-aware policy enforcement
 *
 * Usage:
 *   npm install ws
 *   npx ts-node examples/websocket_server.ts
 */

import WebSocket from 'ws';
import { createServer } from 'http';

// AHP Protocol Types
interface AhpRequest {
  jsonrpc: string;
  id: string;
  method: string;
  params: any;
}

interface AhpNotification {
  jsonrpc: string;
  method: string;
  params: any;
}

interface AhpResponse {
  jsonrpc: string;
  id: string;
  result?: any;
  error?: {
    code: number;
    message: string;
  };
}

interface AhpEvent {
  event_type: string;
  session_id: string;
  agent_id: string;
  timestamp: string;
  depth: number;
  payload: any;
  metadata?: Record<string, any>;
}

interface Decision {
  decision: 'allow' | 'block' | 'modify' | 'defer' | 'escalate';
  reason?: string;
  modified_payload?: any;
  retry_after_ms?: number;
  metadata?: Record<string, any>;
}

// Policy configuration
const DANGEROUS_PATTERNS = [
  /rm\s+-rf\s+\//,
  /dd\s+if=/,
  /mkfs/,
  /:\(\)\{\s*:\|:&\s*\};:/,
  />\s*\/dev\/sda/,
  /wget.*\|\s*sh/,
  /curl.*\|\s*bash/,
];

/**
 * Check if a command is dangerous
 */
function isDangerous(command: string): boolean {
  return DANGEROUS_PATTERNS.some(pattern => pattern.test(command));
}

/**
 * Handle handshake request
 */
function handleHandshake(params: any): any {
  console.log('[HANDSHAKE] Client:', params.agent_info?.framework, params.agent_info?.version);

  return {
    protocol_version: '2.0',
    harness_info: {
      name: 'typescript-websocket-harness',
      version: '1.0.0',
      capabilities: ['pre_action', 'post_action', 'pre_prompt', 'query', 'batch', 'streaming'],
    },
    config: {
      timeout_ms: 10000,
      batch_size: 100,
      max_depth: 10,
    },
  };
}

/**
 * Handle event with depth-aware policy
 */
function handleEvent(event: AhpEvent): Decision {
  const { event_type, depth, payload } = event;

  console.log(`[EVENT] ${event_type} (depth: ${depth}, session: ${event.session_id.substring(0, 8)}...)`);

  if (event_type === 'pre_action') {
    const command = payload?.arguments?.command;

    if (command) {
      // Depth-aware policy: stricter rules for nested agents
      const depthMultiplier = depth === 0 ? 1.0 : depth === 1 ? 1.3 : 1.8;

      if (isDangerous(command)) {
        console.log(`[BLOCK] Dangerous command at depth ${depth}: ${command}`);
        return {
          decision: 'block',
          reason: `Dangerous command detected: ${command}`,
          metadata: {
            policy: 'security',
            depth,
            severity: 'high',
          },
        };
      }

      // Block network access for deeply nested agents
      if (depth > 1 && /curl|wget|nc|telnet|ssh/.test(command)) {
        console.log(`[BLOCK] Network access denied at depth ${depth}`);
        return {
          decision: 'block',
          reason: `Network access not allowed at depth ${depth}`,
          metadata: {
            policy: 'network-isolation',
            depth,
          },
        };
      }

      // Rate limiting for high-depth agents
      if (depth > 2) {
        console.log(`[DEFER] Rate limiting at depth ${depth}`);
        return {
          decision: 'defer',
          retry_after_ms: 1000 * depthMultiplier,
          reason: 'Rate limiting for nested agents',
        };
      }
    }
  }

  return {
    decision: 'allow',
    metadata: {
      depth,
      timestamp: new Date().toISOString(),
    },
  };
}

/**
 * Handle query request
 */
function handleQuery(params: any): any {
  const question = params.payload?.question || '';
  const filePath = params.payload?.file_path;

  console.log(`[QUERY] ${question}`);

  if (question.toLowerCase().includes('delete')) {
    return {
      answer: 'no',
      reason: 'Deletion requires explicit user confirmation',
      alternatives: ['Move to trash', 'Create backup first', 'Mark for review'],
    };
  }

  if (filePath && filePath.includes('important')) {
    return {
      answer: 'no',
      reason: 'File is marked as important',
      alternatives: ['Review file contents', 'Check file metadata'],
    };
  }

  return {
    answer: 'yes',
    reason: 'No concerns detected',
  };
}

/**
 * Handle batch request
 */
function handleBatch(params: any): any {
  const events: AhpEvent[] = params.events || [];
  console.log(`[BATCH] Processing ${events.length} events`);

  const decisions = events.map(event => handleEvent(event));

  const blocked = decisions.filter(d => d.decision === 'block').length;
  const allowed = decisions.filter(d => d.decision === 'allow').length;

  console.log(`[BATCH] Results: ${allowed} allowed, ${blocked} blocked`);

  return { decisions };
}

/**
 * Handle WebSocket message
 */
function handleMessage(ws: WebSocket, message: string) {
  try {
    const msg = JSON.parse(message);

    // Check if it's a request (has id) or notification (no id)
    if (msg.id) {
      // Request - send response
      const request: AhpRequest = msg;
      let result: any;

      switch (request.method) {
        case 'ahp/handshake':
          result = handleHandshake(request.params);
          break;

        case 'ahp/event':
          result = handleEvent(request.params);
          break;

        case 'ahp/query':
          result = handleQuery(request.params);
          break;

        case 'ahp/batch':
          result = handleBatch(request.params);
          break;

        default:
          ws.send(JSON.stringify({
            jsonrpc: '2.0',
            id: request.id,
            error: {
              code: -32601,
              message: `Method not found: ${request.method}`,
            },
          }));
          return;
      }

      const response: AhpResponse = {
        jsonrpc: '2.0',
        id: request.id,
        result,
      };

      ws.send(JSON.stringify(response));
    } else {
      // Notification - no response needed
      const notification: AhpNotification = msg;
      console.log(`[NOTIFICATION] ${notification.method}`);

      // Handle notification asynchronously
      if (notification.method === 'ahp/event') {
        handleEvent(notification.params);
      }
    }
  } catch (error) {
    console.error('[ERROR] Failed to handle message:', error);
    ws.send(JSON.stringify({
      jsonrpc: '2.0',
      id: null,
      error: {
        code: -32700,
        message: 'Parse error',
      },
    }));
  }
}

/**
 * Start the WebSocket server
 */
function main() {
  const PORT = parseInt(process.env.PORT || '8081');

  // Create HTTP server for WebSocket upgrade
  const server = createServer((req, res) => {
    if (req.url === '/health') {
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ status: 'ok', protocol: 'AHP v2.0', transport: 'WebSocket' }));
    } else {
      res.writeHead(404);
      res.end('Not Found');
    }
  });

  // Create WebSocket server
  const wss = new WebSocket.Server({ server, path: '/ahp' });

  let clientCount = 0;

  wss.on('connection', (ws: WebSocket, req) => {
    const clientId = ++clientCount;
    const clientIp = req.socket.remoteAddress;

    console.log(`\n[CONNECT] Client #${clientId} connected from ${clientIp}`);

    // Set up ping/pong heartbeat
    let isAlive = true;

    ws.on('pong', () => {
      isAlive = true;
    });

    const heartbeat = setInterval(() => {
      if (!isAlive) {
        console.log(`[DISCONNECT] Client #${clientId} timeout`);
        ws.terminate();
        return;
      }

      isAlive = false;
      ws.ping();
    }, 30000); // 30 seconds

    // Handle messages
    ws.on('message', (data: WebSocket.Data) => {
      const message = data.toString();
      handleMessage(ws, message);
    });

    // Handle close
    ws.on('close', () => {
      console.log(`[DISCONNECT] Client #${clientId} disconnected`);
      clearInterval(heartbeat);
    });

    // Handle errors
    ws.on('error', (error) => {
      console.error(`[ERROR] Client #${clientId}:`, error.message);
    });
  });

  // Start server
  server.listen(PORT, () => {
    console.log(`\n🚀 AHP WebSocket Server listening on ws://0.0.0.0:${PORT}/ahp`);
    console.log(`   Health check: http://0.0.0.0:${PORT}/health`);
    console.log(`   Protocol: AHP v2.0`);
    console.log(`   Transport: WebSocket`);
    console.log(`\nPress Ctrl+C to stop\n`);
  });

  // Graceful shutdown
  process.on('SIGINT', () => {
    console.log('\n\n[SHUTDOWN] Closing server...');
    wss.clients.forEach(client => {
      client.close(1000, 'Server shutting down');
    });
    server.close(() => {
      console.log('[SHUTDOWN] Server closed');
      process.exit(0);
    });
  });
}

// Run if executed directly
if (require.main === module) {
  main();
}

export { handleMessage, handleEvent, handleQuery };

File: crates/a3s-ahp/2.4.0/examples/websocket_server.py
#!/usr/bin/env python3
"""
AHP WebSocket Server (Python)

A WebSocket-based AHP harness server with bidirectional streaming support.

Features:
- Persistent connections with low latency
- Concurrent client support
- Ping/Pong heartbeat
- Batch processing
- Depth-aware policy enforcement

Usage:
    pip install websockets
    python examples/websocket_server.py

Or with custom port:
    python examples/websocket_server.py 8081
"""

import asyncio
import json
import re
import sys
from datetime import datetime
from typing import Any, Dict, List, Optional, Set

import websockets
from websockets.server import WebSocketServerProtocol

# AHP Protocol Types
class Decision:
    def __init__(self, decision: str, reason: Optional[str] = None,
                 modified_payload: Any = None, retry_after_ms: Optional[int] = None,
                 metadata: Optional[Dict] = None):
        self.decision = decision
        self.reason = reason
        self.modified_payload = modified_payload
        self.retry_after_ms = retry_after_ms
        self.metadata = metadata or {}

    def to_dict(self) -> Dict:
        result = {"decision": self.decision}
        if self.reason:
            result["reason"] = self.reason
        if self.modified_payload:
            result["modified_payload"] = self.modified_payload
        if self.retry_after_ms:
            result["retry_after_ms"] = self.retry_after_ms
        if self.metadata:
            result["metadata"] = self.metadata
        return result

# Dangerous command patterns
DANGEROUS_PATTERNS = [
    re.compile(r'rm\s+-rf\s+/'),
    re.compile(r'dd\s+if='),
    re.compile(r'mkfs'),
    re.compile(r':\(\)\{\s*:\|:&\s*\};:'),  # Fork bomb
    re.compile(r'>\s*/dev/sda'),
    re.compile(r'wget.*\|\s*sh'),
    re.compile(r'curl.*\|\s*bash'),
]

def is_dangerous(command: str) -> bool:
    """Check if a command is dangerous"""
    return any(pattern.search(command) for pattern in DANGEROUS_PATTERNS)

def handle_handshake(params: Dict) -> Dict:
    """Handle handshake request"""
    agent_info = params.get('agent_info', {})
    framework = agent_info.get('framework', 'unknown')
    version = agent_info.get('version', 'unknown')
    print(f'[HANDSHAKE] Client: {framework} {version}')

    return {
        'protocol_version': '2.0',
        'harness_info': {
            'name': 'python-websocket-harness',
            'version': '1.0.0',
            'capabilities': ['pre_action', 'post_action', 'pre_prompt', 'query', 'batch', 'streaming'],
        },
        'config': {
            'timeout_ms': 10000,
            'batch_size': 100,
            'max_depth': 10,
        },
    }

def handle_event(event: Dict) -> Dict:
    """Handle event with depth-aware policy"""
    event_type = event.get('event_type')
    depth = event.get('depth', 0)
    payload = event.get('payload', {})
    session_id = event.get('session_id', '')

    print(f'[EVENT] {event_type} (depth: {depth}, session: {session_id[:8]}...)')

    if event_type == 'pre_action':
        command = payload.get('arguments', {}).get('command')

        if command:
            # Depth-aware policy: stricter rules for nested agents
            depth_multiplier = 1.0 if depth == 0 else (1.3 if depth == 1 else 1.8)

            if is_dangerous(command):
                print(f'[BLOCK] Dangerous command at depth {depth}: {command}')
                return Decision(
                    decision='block',
                    reason=f'Dangerous command detected: {command}',
                    metadata={
                        'policy': 'security',
                        'depth': depth,
                        'severity': 'high',
                    }
                ).to_dict()

            # Block network access for deeply nested agents
            if depth > 1 and re.search(r'curl|wget|nc|telnet|ssh', command):
                print(f'[BLOCK] Network access denied at depth {depth}')
                return Decision(
                    decision='block',
                    reason=f'Network access not allowed at depth {depth}',
                    metadata={
                        'policy': 'network-isolation',
                        'depth': depth,
                    }
                ).to_dict()

            # Rate limiting for high-depth agents
            if depth > 2:
                print(f'[DEFER] Rate limiting at depth {depth}')
                return Decision(
                    decision='defer',
                    retry_after_ms=int(1000 * depth_multiplier),
                    reason='Rate limiting for nested agents'
                ).to_dict()

    return Decision(
        decision='allow',
        metadata={
            'depth': depth,
            'timestamp': datetime.utcnow().isoformat() + 'Z',
        }
    ).to_dict()

def handle_query(params: Dict) -> Dict:
    """Handle query request"""
    question = params.get('payload', {}).get('question', '')
    file_path = params.get('payload', {}).get('file_path')

    print(f'[QUERY] {question}')

    if 'delete' in question.lower():
        return {
            'answer': 'no',
            'reason': 'Deletion requires explicit user confirmation',
            'alternatives': ['Move to trash', 'Create backup first', 'Mark for review'],
        }

    if file_path and 'important' in file_path:
        return {
            'answer': 'no',
            'reason': 'File is marked as important',
            'alternatives': ['Review file contents', 'Check file metadata'],
        }

    return {
        'answer': 'yes',
        'reason': 'No concerns detected',
    }

def handle_batch(params: Dict) -> Dict:
    """Handle batch request"""
    events = params.get('events', [])
    print(f'[BATCH] Processing {len(events)} events')

    decisions = [handle_event(event) for event in events]

    blocked = sum(1 for d in decisions if d['decision'] == 'block')
    allowed = sum(1 for d in decisions if d['decision'] == 'allow')

    print(f'[BATCH] Results: {allowed} allowed, {blocked} blocked')

    return {'decisions': decisions}

async def handle_message(websocket: WebSocketServerProtocol, message: str):
    """Handle WebSocket message"""
    try:
        msg = json.loads(message)

        # Check if it's a request (has id) or notification (no id)
        if 'id' in msg:
            # Request - send response
            method = msg.get('method')
            params = msg.get('params', {})
            req_id = msg.get('id')

            result = None

            if method == 'ahp/handshake':
                result = handle_handshake(params)
            elif method == 'ahp/event':
                result = handle_event(params)
            elif method == 'ahp/query':
                result = handle_query(params)
            elif method == 'ahp/batch':
                result = handle_batch(params)
            else:
                await websocket.send(json.dumps({
                    'jsonrpc': '2.0',
                    'id': req_id,
                    'error': {
                        'code': -32601,
                        'message': f'Method not found: {method}',
                    },
                }))
                return

            response = {
                'jsonrpc': '2.0',
                'id': req_id,
                'result': result,
            }

            await websocket.send(json.dumps(response))
        else:
            # Notification - no response needed
            method = msg.get('method')
            print(f'[NOTIFICATION] {method}')

            # Handle notification asynchronously
            if method == 'ahp/event':
                handle_event(msg.get('params', {}))

    except json.JSONDecodeError:
        await websocket.send(json.dumps({
            'jsonrpc': '2.0',
            'id': None,
            'error': {
                'code': -32700,
                'message': 'Parse error',
            },
        }))
    except Exception as e:
        print(f'[ERROR] Failed to handle message: {e}', file=sys.stderr)
        await websocket.send(json.dumps({
            'jsonrpc': '2.0',
            'id': None,
            'error': {
                'code': -32603,
                'message': f'Internal error: {str(e)}',
            },
        }))

# Track connected clients
connected_clients: Set[WebSocketServerProtocol] = set()
client_counter = 0

async def handle_client(websocket: WebSocketServerProtocol, path: str):
    """Handle WebSocket client connection"""
    global client_counter
    client_counter += 1
    client_id = client_counter
    client_ip = websocket.remote_address[0] if websocket.remote_address else 'unknown'

    print(f'\n[CONNECT] Client #{client_id} connected from {client_ip}')
    connected_clients.add(websocket)

    try:
        # Heartbeat task
        async def heartbeat():
            try:
                while True:
                    await asyncio.sleep(30)
                    await websocket.ping()
            except asyncio.CancelledError:
                pass
            except Exception:
                pass

        heartbeat_task = asyncio.create_task(heartbeat())

        # Handle messages
        async for message in websocket:
            await handle_message(websocket, message)

    except websockets.exceptions.ConnectionClosed:
        print(f'[DISCONNECT] Client #{client_id} disconnected')
    except Exception as e:
        print(f'[ERROR] Client #{client_id}: {e}', file=sys.stderr)
    finally:
        heartbeat_task.cancel()
        connected_clients.discard(websocket)

async def main():
    """Start the WebSocket server"""
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8081

    print(f'\n🚀 AHP WebSocket Server listening on ws://0.0.0.0:{port}/ahp')
    print(f'   Protocol: AHP v2.0')
    print(f'   Transport: WebSocket')
    print(f'\nPress Ctrl+C to stop\n')

    async with websockets.serve(handle_client, '0.0.0.0', port, subprotocols=['ahp']):
        await asyncio.Future()  # Run forever

if __name__ == '__main__':
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print('\n\n[SHUTDOWN] Server stopped')
        sys.exit(0)
